### PR TITLE
Add option for synchronizing dosbox to the front-end

### PR DESCRIPTION
**Pull request by realnc**
The current method of timer-based scheduling to run the emulation thread (using PIC_AddEvent) results in dosbox and the front-end not running in sync. Dosbox does not render exactly one frame at the correct time so we have to always copy the last rendered frame and keep it around until the front-end requests it. As a result, there is increased input lag of up to 1 frame.

We fix this by doing a context switch to the emulator and back to the front-end exactly once every time the front-end requests a frame. We add a small change to dosbox to make it keep presenting frames even when there were no changes in the framebuffer. After presenting the frame, we switch back to the front-end. For this to work reliably, we put dosbox in fast-forward mode at startup.

As a result, there is no more delay between dosbox finishing drawing the frame and uploading it to the front-end, minimizing input lag. Also, for near-60FPS games at least, the front-end can now successfully reclock the core to match the screen refresh rate, so there shouldn't be any stutter or judder.

There is a drawback to this though: the "auto" and "max" cycles settings do not work correctly anymore. This is a side-effect of running dosbox at an unlocked speed. As a result, this mode is disabled by default and needs to be enabled in the core options.

To make it easier for users to test the three different timing modes, the "switch refresh rate" option has been removed. Now there's only one "Core: Timing method" option with three settings:

  * Independent, run frontend at 60FPS
  * Independent, run frontend at emulated refresh rate
  * Synchronize to frontend

Also, this setting can be changed at runtime and will take effect immediately without the need for a restart.

The option takes effect even if core options are disabled, since this is more of a libretro-specific setting rather than a dosbox one.

-------------

This should probably get some more testing before being merged. Works as advertised for me for all games I tested. Games respond 1 frame sooner when testing them with RA's frame advance function (`k` key), and the mouse cursor in point&click games is now 1 frame closer to the native mouse cursor rather than lagging two frames behind. I think input lag now matches stand-alone dosbox.

But I'm not 100% sure that `RENDER_Halt()` is the only place where a context switch is needed. So try as many games as you can and see if the UI locks up with any of them.

### CODE REVIEW

**Comment by fr500** [comment]
First of all thank you for this.
One thing is that Fast Forward should now work without this:

```C
    environ_cb(RETRO_ENVIRONMENT_GET_FASTFORWARDING, &fast_forward);
    DOSBOX_UnlockSpeed(fast_forward);
```

**Comment by fr500** [comment]
I tested this and sadly the experience for me is horrible.
I'm testing this on a non g-sync 60hz monitor. I tried a simplistic game (KEEN 5).

Without refresh rate switching the game is now slow and jittery (before it would skip frames but play at the correct speed, felt a lot smoother)

With refresh rate switching it plays at the correct speed. fast forward works, slow motion works, but there is an odd judder here and there ever few seconds, major judder, and obvious tearing.

I know the way it's working now is far from perfect but somehow it feels smooth, even when running 70hz on a vsynced 60hz screen. The judder now is far more noticeable for me.

Bottom line is that the old mode has less drawbacks for my personal use case. For non-VRR users I feel like it would be the same.

Maybe we can keep both behaviors in-place? I wouldn't be against that at all.

**Comment by fr500** [comment]
I tried another game (Jazz), I basically pressed both left and right at the same time so jazz would walk slowly. There is that judder again every few seconds. It's buttery smooth (better than before) the rest of the time, but there is this odd judder every few frames.

**Comment by realnc** [comment]
Since dosbox doesn't support reclocking, refresh rate switching is needed. Stand-alone dosbox should behave the same way. It does render at the emulated refresh rate.

How does stand-alone dosbox behave for you? This PR should pretty much make the core match stand-alone dosbox behavior.

**Comment by fr500** [comment]
more than judder, it's as if some frames are updating back and forth when the judder occurs, as if it would rewind a bit and forward again for some reason whenever it happens.

**Comment by realnc** [comment]
I think the stutter occurs because dosbox is still using its own timing, so each call to `retro_run()` takes too long, not giving the front-end enough time before the next call.

I now put dosbox in fast-forward mode at startup. This should allow RA to reclock near-60FPS games to the real screen refresh rate, so it should be smooth now. Hopefully. 70FPS are another matter though... See:

https://forums.libretro.com/t/retroarch-runs-cores-without-vsync-if-they-want-a-higher-fps-than-60/24910

**Comment by realnc** [comment]
I just realized that this breaks the "max" setting for cycles :-/ So this option requires a fixed cycle amount. Not sure if this can be avoided.

**Comment by realnc** [comment]
Hm. Never mind. Max cycles is already broken. It doesn't have anything to do with this patch.

**Comment by fr500** [comment]
@realnc cycles max is working fine here, just tested with quake, auto is smooth, fixed default is unbearable, max is smooth again

**Comment by fr500** [comment]
Even with your patch (but with timing independent) max and auto work fine.
With your patch though neither work and you need fixed cycles.

**Comment by realnc** [comment]
I guess it's just on Linux that "auto" and "max" are broken.

I updated the PR. I removed the "refresh rate change" option and added it as another core timing option. Also, you can change timing mode at runtime now without the need for a restart.

This PR can be merged now, unless somebody finds regressions.

**Comment by fr500** [comment]
New one doesn't work like... at all for me
Only option that works properly is independent 60.

I liked the earlier approach better too (less messing around with dosbox code)
That said if it worked properly I'd merge this as-is

**Comment by realnc** [comment]
I've been staring at the code for hours. Can't find anything wrong and I have no clue why it doesn't work for you :-/

I built the core for 32-bit Windows (don't have a 64-bit build environment handy right now), installed RA 1.8.1 and tried the core. Works fine:

http://83.212.109.87/~realnc/tmp/dosbox_svn_libretro.dll

**Comment by fr500** [comment]
seems to work fine on my end now....
weird, I'll merge it, the old behavior is still in place anyway
